haystack/preview/components/prompt/models/base.py

```diff
@@ -0,0 +1,103 @@
+from typing import Type, Optional, Dict, Any, List
+
+import logging
+
+from canals import _find_decorated_classes
+
+
+logger = logging.getLogger(__name__)
+
+
+class PromptModelError(Exception):
+    pass
+
+
+def prompt_model(class_):
+    """
+    Prompt model implementations execute a prompt on an underlying model.
+
+    The implementation can be a simple invocation on the underlying model running in a local runtime, or
+    could be even remote, for example, a call to a remote API endpoint.
+
+    `def __init__(self, model_name_or_path: str, **kwargs)`
+
+        Creates a new PromptModelInvocationLayer instance.
+
+        :param model_name_or_path: The name or path of the underlying model.
+        :param kwargs: Additional keyword arguments passed to the underlying model.
+
+    `def invoke(self, *args, **kwargs)`
+
+        It takes a prompt and returns a list of generated text using the underlying model.
+        :return: A list of generated text.
+
+    `def supports(cls, model_name_or_path: str, **kwargs) -> bool`
+
+        Checks if the given model is supported by this invocation layer.
+
+        :param model_name_or_path: The name or path of the model.
+        :param kwargs: additional keyword arguments passed to the underlying model which might be used to determine
+        if the model is supported.
+        :return: True if this invocation layer supports the model, False otherwise.
+
+    """
+    logger.debug("Registering %s as a prompt model implementation", class_)
+
+    # '__haystack_prompt_model__' is used to distinguish prompt model implementations from regular classes.
+    # Its value is set to the desired implementation name: normally it is the class name, but it can technically be customized.
+    class_.__haystack_prompt_model__ = class_.__name__
+
+    # Check for invoke()
+    if not hasattr(class_, "invoke"):
+        # TODO check the node signature too
+        raise PromptModelError(
+            "Prompt model implementations must have an 'invoke()' method. See the docs for more information."
+        )
+
+    # Check for supports()
+    if not hasattr(class_, "supports"):
+        # TODO check the node signature too
+        raise PromptModelError(
+            "Prompt model implementations  must have a 'supports()' method. See the docs for more information."
+        )
+
+    return class_
+
+
+def get_model(
+    model_name_or_path: str,
+    model_implementation: Optional[Type] = None,
+    model_params: Optional[Dict[str, Any]] = None,
+    modules_to_search: Optional[List[str]] = None,
```

Comment:
What's the use-case for passing modules instead of passing Class or Object? Why should Haystack take care of searching through the whole module?

What happens if there are multiple matching "implementations" available in the same module supporting the same model? (i.e., CustomSeq2Seq and CustomPromptModel both supporting "lfqa" model)